File: Crio/crio_programming_interview_problems-master/WordLadder/WordLadder.py
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
def isValidTranformation(a ,b ):
    difference = 0;
    for i in range(len(a)):
        if a[i] != b[i]:
            difference+=1
    return difference == 1;
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def ladderLength(beginWord, endWord, wordList):
    # CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    wordList.append(beginWord)
    vec = [[] for _ in range(len(wordList))]
    destination = -1
    for i in range(len(wordList)):
        for j in range(i+1,len(wordList)):
            if isValidTranformation(wordList[i],wordList[j]):
                vec[i].append(j)
                vec[j].append(i)

        if wordList[i]==endWord :
            destination =i

    q = list()
    level=[0 for _ in range(len(wordList))]
    q.append(len(wordList)-1)
    level[len(wordList)-1]=1
    while len(q) > 0:
        front = q.pop(0)
        for i in range(len(vec[front])):
            if level[vec[front][i]]==0:
                q.append(vec[front][i])
                level[vec[front][i]]=level[front]+1

    if destination == -1:
        return 0
    elif level[destination] == 0:
        return 0
    else:
        return level[destination]
    # CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

if __name__ == '__main__':
    beginWord = input()
    endWord = input()
    n = int(input())
    wordList = []
    if n > 0:
        wordList = input().split()
    result = ladderLength(beginWord, endWord, wordList)
    print(result)

File: Crio/crio_programming_interview_problems-master/WordLadder/WordLadder.cpp
#include <bits/stdc++.h>
using namespace std;

class WordLadder {
  public:
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        bool isValidTranformation(string a, string b)
        {
            int difference = 0;
            for (int i = 0; i < a.length(); ++i)
            {
                if (a[i] != b[i]) {
                    difference++;
                }
            }
            return difference == 1;
        }
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
        int ladderLength(string beginWord, string endWord,vector<string> wordList) 
        {
            // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
            wordList.push_back(beginWord);
            vector<vector <int> > vec(wordList.size());
            int destination = -1;
            for(int i=0;i<wordList.size();i++)
            {
                for(int j=i+1;j<wordList.size();j++)
                {
                    if(isValidTranformation(wordList[i],wordList[j]))
                    {
                        vec[i].push_back(j);
                        vec[j].push_back(i);
                    }
                }
                if(wordList[i]==endWord)
                    destination =i;
            }


            queue<int> q;
            int level[wordList.size()]={0};
            q.push(wordList.size()-1);
            level[wordList.size()-1]=1;
            while(!q.empty())
            {
                int front = q.front();
                q.pop();
                for(int i=0;i<vec[front].size();i++)
                {
                    if(level[vec[front][i]]==0)
                    {
                        q.push(vec[front][i]);
                        level[vec[front][i]]=level[front]+1;
                    }
                }
            }

            
            if(destination == -1)
                return 0;
            else if(level[destination]==0)
                return 0;
            else
                return level[destination];

            // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
        }
};

int main() {
    string beginWord, endWord;
    cin >> beginWord >> endWord;
    int n;
    cin >> n;
    vector<string> wordList;
    for(int i =0;i<n;i++)
    {
        string s;
        cin>>s;
        wordList.push_back(s);
    }
    int result = WordLadder().ladderLength(beginWord, endWord, wordList);
    cout << result;
    return 0;
}
